R/driver-snowflake.R

```diff
@@ -5,6 +5,58 @@ NULL
 #' @rdname DBI-classes
 setClass("Snowflake", contains = "OdbcConnection")
 
+#' Connecting to Snowflake via ODBC
+#'
+#' @description
+#' ## `odbcConnectionColumns()`
+#'
+#' If the catalog, or the schema arguments are NULL, attempt to infer
+#' by querying for CURRENT_DATABASE() and CURRENT_SCHEMA().  We do this
+#' to aid with performance, as the SQLColumns method is more performant
+#' when restricted to a particular DB/schema.
+#' @inheritParams DBI::dbListFields
+#' @rdname Snowflake
+#' @usage NULL
+setMethod("odbcConnectionColumns_", c("Snowflake", "character"),
+  function(conn,
+           name,
+           ...,
+           catalog_name = NULL,
+           schema_name = NULL,
+           column_name = NULL,
+           exact = FALSE) {
+    if(is.null(catalog_name) || is.null(schema_name)) {
+      res <- dbGetQuery(conn, "SELECT CURRENT_DATABASE() AS CAT, CURRENT_SCHEMA() AS SCH")
+      if(is.null(catalog_name) && !is.na(res$CAT)) {
+        catalog_name <- res$CAT
+      }
+      if(is.null(schema_name) && !is.na(res$SCH)) {
+        schema_name <- res$SCH
+      }
+    }
+    callNextMethod(conn = conn, name = name, ..., catalog_name = catalog_name,
+      schema_name = schema_name, column_name = column_name, exact = exact)
+  }
```

Comment:
I think you only need to supply the arguments that have changed?

```suggestion
    callNextMethod(catalog_name = catalog_name, schema_name = schema_name)
  }
```

Reply:
Hey Hadley -  good thought, but I don't think this works ( or i screwed something up when trying ).

Reply:
Let me take a look.

Reply:
Oh yeah, you're correct. I had misunderstood the docs for `callNextMethod()`:

``` r
setClass("A")

1. setClass("B", contains = "A")

setGeneric("foo", function(x, a, b) standardGeneric("foo"))
#> [1] "foo"
setMethod("foo", "A", function(x, a, b) {
  c(a = a, b = b)
})
setMethod("foo", "B", function(x, a, b) {
  callNextMethod()
})
foo(new("B"), a = 1, b = 1)
#> a b 
#> 1 1

setMethod("foo", "B", function(x, a, b) {
  callNextMethod(b = 2)
})
foo(new("B"), a = 1, b = 1)
#> Error in .nextMethod(b = 2): argument "a" is missing, with no default
```

<sup>Created on 2024-03-05 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>